

<!DOCTYPE html>


<html lang="en" data-content_root="" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

    <title>Exercise 3: Spatial architecture of tumor microenvironment with SCIMAP &#8212; Spatial Proteomics Workshop 2026</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="_static/styles/theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="_static/styles/bootstrap.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="_static/styles/pydata-sphinx-theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />

  
  <link href="_static/vendor/fontawesome/6.5.2/css/all.min.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.5.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.5.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.5.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="_static/pygments.css" />
    <link rel="stylesheet" href="_static/styles/sphinx-book-theme.css?digest=14f4ca6b54d191a8c7657f6c759bf11a5fb86285" type="text/css" />
    <link rel="stylesheet" type="text/css" href="_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css" />
    <link rel="stylesheet" type="text/css" href="_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="_static/design-style.4045f2051d55cab465a707391d5b2007.min.css" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b" />
<link rel="preload" as="script" href="_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b" />
  <script src="_static/vendor/fontawesome/6.5.2/js/all.min.js?digest=dfe6caa3a7d634c4db9b"></script>

    <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/_sphinx_javascript_frameworks_compat.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/clipboard.min.js"></script>
    <script src="_static/copybutton.js"></script>
    <script src="_static/scripts/sphinx-book-theme.js?digest=5a5c038af52cf7bc1a1ec88eea08e6366ee68824"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="_static/togglebutton.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="_static/design-tabs.js"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"
const thebe_selector = ".thebe,.cell"
const thebe_selector_input = "pre"
const thebe_selector_output = ".output, .cell_output"
</script>
    <script async="async" src="_static/sphinx-thebe.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'exercise3';</script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="prev" title="Exercise 2: Biological interpretation &amp; cell phenotyping" href="exercise2.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-primary-sidebar-checkbox"/>
  <label class="overlay overlay-primary" for="pst-primary-sidebar-checkbox"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-secondary-sidebar-checkbox"/>
  <label class="overlay overlay-secondary" for="pst-secondary-sidebar-checkbox"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search this book..."
         aria-label="Search this book..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">

  
    
  

<a class="navbar-brand logo" href="index.html">
  
  
  
  
  
  
    <p class="title logo__title">Spatial Proteomics Workshop 2026</p>
  
</a></div>
        <div class="sidebar-primary-item"><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="index.html">
                    Spatial Proteomics: From Pixels to Phenotypes
                </a>
            </li>
        </ul>
        <ul class="current nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="exercise1.html">Exercise 1: Technical foundations of digital pathology</a></li>
<li class="toctree-l1"><a class="reference internal" href="exercise2.html">Exercise 2: Biological interpretation &amp; cell phenotyping</a></li>
<li class="toctree-l1 current active"><a class="current reference internal" href="#">Exercise 3: Spatial architecture of tumor microenvironment with SCIMAP</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><label class="sidebar-toggle primary-toggle btn btn-sm" for="__primary" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</label></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="_sources/exercise3.ipynb" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.ipynb</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>



<script>
document.write(`
  <button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="theme-switch fa-solid fa-sun fa-lg" data-mode="light"></i>
    <i class="theme-switch fa-solid fa-moon fa-lg" data-mode="dark"></i>
    <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"></i>
  </button>
`);
</script>


<script>
document.write(`
  <button class="btn btn-sm pst-navbar-icon search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
  </button>
`);
</script>
<label class="sidebar-toggle secondary-toggle btn btn-sm" for="__secondary"title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</label>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>Exercise 3: Spatial architecture of tumor microenvironment with SCIMAP</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#clinical-context">Clinical Context</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#what-you-will-discover">What You Will Discover</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#learning-outcomes">Learning outcomes</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#workflow-overview">Workflow Overview</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#step-1-load-the-data">Step 1: Load the Data</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#step-2-visualise-raw-marker-expression">Step 2: Visualise Raw Marker Expression</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#marker-reference-table">Marker Reference Table</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#step-3-define-cell-identities">Step 3: Define Cell Identities</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#the-rules">The Rules</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#run-the-classification">Run the Classification</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#step-4-quality-control">Step 4: Quality Control</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#step-5-measuring-the-social-distance-between-cell-types">Step 5: Measuring the “Social Distance” Between Cell Types</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#why-does-this-matter">Why does this matter?</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#step-6-mapping-recurrent-cell-neighbourhoods-rcns">Step 6: Mapping Recurrent Cell Neighbourhoods (RCNs)</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#the-rcn-approach-in-three-steps">The RCN approach in three steps:</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#finding-the-optimal-number-of-neighbourhoods">Finding the Optimal Number of Neighbourhoods</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#interpreting-the-neighbourhoods">Interpreting the Neighbourhoods</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#assign-biological-names-to-clusters">Assign Biological Names to Clusters</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#summary-conclusions">Summary &amp; Conclusions</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#key-takeaways">Key Takeaways</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#going-further">Going Further</a></li>
</ul>
            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section class="tex2jax_ignore mathjax_ignore" id="exercise-3-spatial-architecture-of-tumor-microenvironment-with-scimap">
<h1>Exercise 3: Spatial architecture of tumor microenvironment with SCIMAP<a class="headerlink" href="#exercise-3-spatial-architecture-of-tumor-microenvironment-with-scimap" title="Permalink to this heading">#</a></h1>
<p><strong>Objective:</strong> Use Scimap to determine whether immune cells are successfully infiltrating the tumor — or getting “trapped” at the interface.</p>
<hr class="docutils" />
<section id="clinical-context">
<h2>Clinical Context<a class="headerlink" href="#clinical-context" title="Permalink to this heading">#</a></h2>
<p>This exercise uses data from a study of <strong>high-grade serous ovarian cancer (HGSOC)</strong> — one of the most lethal gynaecological cancers, largely because most patients eventually develop resistance to standard platinum-based chemotherapy.</p>
<p>Two major treatment strategies exist, and they attract very different immune landscapes:</p>
<div class="pst-scrollable-table-container"><table class="table">
<thead>
<tr class="row-odd"><th class="head"><p>Strategy</p></th>
<th class="head"><p>Abbreviation</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>Primary Debulking Surgery</p></td>
<td><p><strong>PDS</strong></p></td>
<td><p>Surgery first, then chemotherapy</p></td>
</tr>
<tr class="row-odd"><td><p>Interval Debulking Surgery</p></td>
<td><p><strong>IDS</strong></p></td>
<td><p>Chemotherapy first to shrink the tumour, then surgery</p></td>
</tr>
</tbody>
</table>
</div>
<p>Patients are also stratified by their <strong>homologous recombination (HR) status</strong> — a measure of DNA repair capacity:</p>
<ul class="simple">
<li><p><strong>BRCALoss</strong> — BRCA1/2 mutations impair DNA repair, making tumours more sensitive to certain therapies and more immunogenic</p></li>
<li><p><strong>HRP</strong> (HR-proficient) — intact DNA repair, generally associated with poorer immunotherapy response</p></li>
</ul>
<p>In the study (<a class="reference external" href="https://doi.org/10.1158/2159-8290.CD-25-1492">Perez-Villatoro et al., 2026</a>), different treatment–genotype combinations produced strikingly different <strong>tumour–stroma interfaces (TSI)</strong> — the boundary zone where immune cells either successfully invade the tumour or get blocked.</p>
</section>
<hr class="docutils" />
<section id="what-you-will-discover">
<h2>What You Will Discover<a class="headerlink" href="#what-you-will-discover" title="Permalink to this heading">#</a></h2>
<p>We will compare two TMA cores that represent opposite ends of the immune infiltration spectrum:</p>
<div class="pst-scrollable-table-container"><table class="table">
<thead>
<tr class="row-odd"><th class="head"><p>Core</p></th>
<th class="head"><p>Patient Group</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><strong>Core 84</strong></p></td>
<td><p>BRCALoss + PDS</p></td>
</tr>
<tr class="row-odd"><td><p><strong>Core 42</strong></p></td>
<td><p>HRP + IDS</p></td>
</tr>
</tbody>
</table>
</div>
<blockquote>
<div><p><strong>Key question:</strong> Can you detect the differences in spatial architecture of tumor microenvironment computationally — using only the spatial coordinates and protein expression of individual cells?</p>
</div></blockquote>
</section>
<hr class="docutils" />
<section id="learning-outcomes">
<h2>Learning outcomes<a class="headerlink" href="#learning-outcomes" title="Permalink to this heading">#</a></h2>
<p>By the end of this exercise you will be able to:</p>
<ol class="arabic simple">
<li><p>Visualise raw marker expression across a tissue section.</p></li>
<li><p>Build a rule-based cell phenotyping workflow.</p></li>
<li><p>Quantify the physical proximity of immune cells to tumour cells.</p></li>
<li><p>Identify <strong>Recurrent Cell Neighbourhoods (RCNs)</strong> and interpret their biological meaning.</p></li>
</ol>
</section>
<hr class="docutils" />
<section id="workflow-overview">
<h2>Workflow Overview<a class="headerlink" href="#workflow-overview" title="Permalink to this heading">#</a></h2>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>Step 1 → Load Data
Step 2 → Visualise Raw Markers
Step 3 → Phenotype Cells
Step 4 → Quality Control
Step 5 → Measure Spatial Distances
Step 6 → Map Cell Neighbourhoods
</pre></div>
</div>
</section>
<section id="step-1-load-the-data">
<h2>Step 1: Load the Data<a class="headerlink" href="#step-1-load-the-data" title="Permalink to this heading">#</a></h2>
<p>We start by importing our libraries and loading the dataset. The H5AD format (used by AnnData/Scanpy) stores:</p>
<ul class="simple">
<li><p><strong><code class="docutils literal notranslate"><span class="pre">adata.X</span></code></strong> — the protein expression matrix (cells × markers)</p></li>
<li><p><strong><code class="docutils literal notranslate"><span class="pre">adata.obs</span></code></strong> — per-cell metadata (image ID, coordinates, phenotype labels)</p></li>
<li><p><strong><code class="docutils literal notranslate"><span class="pre">adata.obsm['spatial']</span></code></strong> — the X, Y pixel coordinates of each cell in the tissue image</p></li>
</ul>
<p>After loading, we split the combined dataset into the two TMA cores for independent processing.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">scimap</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">sm</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">scanpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">sc</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">anndata</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">ad</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pd</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">matplotlib.pyplot</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">plt</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">matplotlib.colors</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">mcolors</span>

<span class="c1"># Load the combined dataset</span>
<span class="n">adata</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read_h5ad</span><span class="p">(</span><span class="s1">&#39;single_cell_TMA_dataset.h5ad&#39;</span><span class="p">)</span>

<span class="c1"># Split into the two TMA cores for independent processing</span>
<span class="n">tma8</span> <span class="o">=</span> <span class="n">adata</span><span class="p">[</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;imageid&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;TMA_8_core84&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="n">tma1</span> <span class="o">=</span> <span class="n">adata</span><span class="p">[</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;imageid&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;TMA_1_core42&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>

<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Loaded </span><span class="si">{</span><span class="n">tma8</span><span class="o">.</span><span class="n">n_obs</span><span class="si">:</span><span class="s2">,</span><span class="si">}</span><span class="s2"> cells for Core 84 and </span><span class="si">{</span><span class="n">tma1</span><span class="o">.</span><span class="n">n_obs</span><span class="si">:</span><span class="s2">,</span><span class="si">}</span><span class="s2"> cells for Core 42.&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">Dataset schema:&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;  Markers : </span><span class="si">{</span><span class="nb">list</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">var_names</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;  Metadata columns : </span><span class="si">{</span><span class="nb">list</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;  Spatial coordinates stored in : adata.obsm[&#39;spatial&#39;]&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output traceback highlight-ipythontb notranslate"><div class="highlight"><pre><span></span><span class="gt">---------------------------------------------------------------------------</span>
<span class="ne">ModuleNotFoundError</span><span class="g g-Whitespace">                       </span>Traceback (most recent call last)
<span class="n">Cell</span> <span class="n">In</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">line</span> <span class="mi">1</span>
<span class="ne">----&gt; </span><span class="mi">1</span> <span class="kn">import</span><span class="w"> </span><span class="nn">scimap</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">sm</span>
<span class="g g-Whitespace">      </span><span class="mi">2</span> <span class="kn">import</span><span class="w"> </span><span class="nn">scanpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">sc</span>
<span class="g g-Whitespace">      </span><span class="mi">3</span> <span class="kn">import</span><span class="w"> </span><span class="nn">anndata</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">ad</span>

<span class="ne">ModuleNotFoundError</span>: No module named &#39;scimap&#39;
</pre></div>
</div>
</div>
</div>
</section>
<section id="step-2-visualise-raw-marker-expression">
<h2>Step 2: Visualise Raw Marker Expression<a class="headerlink" href="#step-2-visualise-raw-marker-expression" title="Permalink to this heading">#</a></h2>
<p>Before calculating anything, we need to build an intuition for the data. We create a <strong>Marker Atlas</strong> — a spatial map of each protein’s expression across the tissue — for both cores side by side.</p>
<p>This serves two purposes:</p>
<ol class="arabic simple">
<li><p>It lets us visually confirm which regions of the tissue express which markers.</p></li>
<li><p>It guides us in writing the phenotyping rules in the next step.</p></li>
</ol>
<blockquote>
<div><p>Generating these plots may take a few seconds.</p>
</div></blockquote>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Define the markers of interest for this exercise</span>
<span class="n">marker_cols</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;CK7&#39;</span><span class="p">,</span> <span class="s1">&#39;ECadherin&#39;</span><span class="p">,</span> <span class="s1">&#39;aSMA&#39;</span><span class="p">,</span> <span class="s1">&#39;Vimentin&#39;</span><span class="p">,</span> <span class="s1">&#39;CD3d&#39;</span><span class="p">,</span> <span class="s1">&#39;CD11c&#39;</span><span class="p">]</span>

<span class="c1"># Custom blue-teal-dark colormap for expression intensity</span>
<span class="n">custom_cmap</span> <span class="o">=</span> <span class="n">mcolors</span><span class="o">.</span><span class="n">LinearSegmentedColormap</span><span class="o">.</span><span class="n">from_list</span><span class="p">(</span>
    <span class="s2">&quot;white_teal_dark&quot;</span><span class="p">,</span> <span class="p">[</span><span class="s2">&quot;#FFFFFF&quot;</span><span class="p">,</span> <span class="s2">&quot;#508791&quot;</span><span class="p">,</span> <span class="s2">&quot;#1D3033&quot;</span><span class="p">]</span>
<span class="p">)</span>

<span class="c1"># --- Marker Atlas: Core 84 ---</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Generating Marker Atlas for Core 84...&quot;</span><span class="p">)</span>
<span class="n">sm</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">spatial_scatterPlot</span><span class="p">(</span>
    <span class="n">tma8</span><span class="p">,</span> <span class="n">colorBy</span><span class="o">=</span><span class="n">marker_cols</span><span class="p">,</span> <span class="n">s</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="n">ncols</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span>
    <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="mi">12</span><span class="p">),</span> <span class="n">cmap</span><span class="o">=</span><span class="n">custom_cmap</span><span class="p">,</span> <span class="n">plotLegend</span><span class="o">=</span><span class="kc">True</span>
<span class="p">)</span>

<span class="c1"># --- Marker Atlas: Core 42 ---</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">Generating Marker Atlas for Core 42...&quot;</span><span class="p">)</span>
<span class="n">sm</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">spatial_scatterPlot</span><span class="p">(</span>
    <span class="n">tma1</span><span class="p">,</span> <span class="n">colorBy</span><span class="o">=</span><span class="n">marker_cols</span><span class="p">,</span> <span class="n">s</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="n">ncols</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span>
    <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="mi">12</span><span class="p">),</span> <span class="n">cmap</span><span class="o">=</span><span class="n">custom_cmap</span><span class="p">,</span> <span class="n">plotLegend</span><span class="o">=</span><span class="kc">True</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<section id="marker-reference-table">
<h3>Marker Reference Table<a class="headerlink" href="#marker-reference-table" title="Permalink to this heading">#</a></h3>
<div class="pst-scrollable-table-container"><table class="table">
<thead>
<tr class="row-odd"><th class="head"><p>Marker</p></th>
<th class="head"><p>Cell Type</p></th>
<th class="head"><p>Biological Role</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><strong>CK7</strong></p></td>
<td><p>Epithelial / Cancer</p></td>
<td><p>Cytokeratin filament protein that delineates primary tumour nests</p></td>
</tr>
<tr class="row-odd"><td><p><strong>E-Cadherin</strong></p></td>
<td><p>Epithelial / Cancer</p></td>
<td><p>Calcium-dependent cell–cell adhesion molecule maintaining epithelial integrity</p></td>
</tr>
<tr class="row-even"><td><p><strong>αSMA</strong></p></td>
<td><p>Stromal / Fibroblast</p></td>
<td><p>Actin isoform expressed in activated cancer-associated fibroblasts</p></td>
</tr>
<tr class="row-odd"><td><p><strong>Vimentin</strong></p></td>
<td><p>Stromal / Mesenchymal</p></td>
<td><p>Intermediate filament broadly marking non-epithelial (mesenchymal) cells</p></td>
</tr>
<tr class="row-even"><td><p><strong>CD3d</strong></p></td>
<td><p>Lymphoid / T cell</p></td>
<td><p>Core component of the T-cell receptor signalling complex</p></td>
</tr>
<tr class="row-odd"><td><p><strong>CD11c</strong></p></td>
<td><p>Myeloid / Dendritic cell</p></td>
<td><p>Integrin expressed at high levels on myeloid cells (DCs, macrophages)</p></td>
</tr>
</tbody>
</table>
</div>
<div class="tip admonition">
<p class="admonition-title">QUESTIONS:</p>
<p>Look at the Marker Atlas for both cores and consider:</p>
<ol class="arabic simple">
<li><p>Which markers show clearly distinct spatial boundaries, and which appear diffuse?</p></li>
<li><p>Based on raw intensity alone, which markers do you expect will be <strong>easy</strong> vs <strong>challenging</strong> to threshold into positives/negatives? Why?</p></li>
<li><p>Do the two cores look similar or different in terms of marker distribution? What does this hint at biologically?</p></li>
</ol>
</div>
</section>
</section>
<section id="step-3-define-cell-identities">
<h2>Step 3: Define Cell Identities<a class="headerlink" href="#step-3-define-cell-identities" title="Permalink to this heading">#</a></h2>
<p>Now that we’ve seen the raw markers, we need to teach the algorithm how to recognise each cell type. We use a <strong>rule-based phenotyping</strong> approach, where each cell type is defined by the presence (<code class="docutils literal notranslate"><span class="pre">pos</span></code>) or absence (<code class="docutils literal notranslate"><span class="pre">NaN</span></code>) of specific markers.</p>
<section id="the-rules">
<h3>The Rules<a class="headerlink" href="#the-rules" title="Permalink to this heading">#</a></h3>
<div class="pst-scrollable-table-container"><table class="table">
<thead>
<tr class="row-odd"><th class="head"><p>Phenotype</p></th>
<th class="head"><p>Required Markers</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><strong>Cancer</strong></p></td>
<td><p>CK7 <strong>AND</strong> ECadherin must be positive</p></td>
</tr>
<tr class="row-odd"><td><p><strong>Stromal</strong></p></td>
<td><p>αSMA <strong>AND</strong> Vimentin must be positive</p></td>
</tr>
<tr class="row-even"><td><p><strong>Myeloid</strong></p></td>
<td><p>CD11c must be positive</p></td>
</tr>
<tr class="row-odd"><td><p><strong>Lymphoid</strong></p></td>
<td><p>CD3d must be positive</p></td>
</tr>
</tbody>
</table>
</div>
<div class="admonition-task admonition">
<p class="admonition-title"><strong>Task:</strong></p>
<p>Fill in the <code class="docutils literal notranslate"><span class="pre">None</span></code> values in the <code class="docutils literal notranslate"><span class="pre">data</span></code> list below with <code class="docutils literal notranslate"><span class="pre">'pos'</span></code> where a marker must be present for each phenotype. Leave <code class="docutils literal notranslate"><span class="pre">None</span></code> where the marker should be absent or irrelevant.</p>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Column order: Global | Phenotype | CK7 | ECadherin | aSMA | Vimentin | CD11c | CD3d</span>
<span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;Global&#39;</span><span class="p">,</span> <span class="s1">&#39;Phenotype&#39;</span><span class="p">,</span> <span class="s1">&#39;CK7&#39;</span><span class="p">,</span> <span class="s1">&#39;ECadherin&#39;</span><span class="p">,</span> <span class="s1">&#39;aSMA&#39;</span><span class="p">,</span> <span class="s1">&#39;Vimentin&#39;</span><span class="p">,</span> <span class="s1">&#39;CD11c&#39;</span><span class="p">,</span> <span class="s1">&#39;CD3d&#39;</span><span class="p">]</span>

<span class="c1"># TODO: Replace None with &#39;pos&#39; where a marker must be expressed for that phenotype</span>
<span class="n">data</span> <span class="o">=</span> <span class="p">[</span>
    <span class="c1">#  Global    Phenotype    CK7    ECad   aSMA   Vim    CD11c  CD3d</span>
    <span class="p">[</span><span class="s1">&#39;all&#39;</span><span class="p">,</span>    <span class="s1">&#39;Cancer&#39;</span><span class="p">,</span>    <span class="kc">None</span><span class="p">,</span>  <span class="kc">None</span><span class="p">,</span>  <span class="kc">None</span><span class="p">,</span>  <span class="kc">None</span><span class="p">,</span>  <span class="kc">None</span><span class="p">,</span>  <span class="kc">None</span><span class="p">],</span>
    <span class="p">[</span><span class="s1">&#39;all&#39;</span><span class="p">,</span>    <span class="s1">&#39;Stromal&#39;</span><span class="p">,</span>   <span class="kc">None</span><span class="p">,</span>  <span class="kc">None</span><span class="p">,</span>  <span class="kc">None</span><span class="p">,</span>  <span class="kc">None</span><span class="p">,</span>  <span class="kc">None</span><span class="p">,</span>  <span class="kc">None</span><span class="p">],</span>
    <span class="p">[</span><span class="s1">&#39;all&#39;</span><span class="p">,</span>    <span class="s1">&#39;Myeloid&#39;</span><span class="p">,</span>   <span class="kc">None</span><span class="p">,</span>  <span class="kc">None</span><span class="p">,</span>  <span class="kc">None</span><span class="p">,</span>  <span class="kc">None</span><span class="p">,</span>  <span class="kc">None</span><span class="p">,</span>  <span class="kc">None</span><span class="p">],</span>
    <span class="p">[</span><span class="s1">&#39;all&#39;</span><span class="p">,</span>    <span class="s1">&#39;Lymphoid&#39;</span><span class="p">,</span>  <span class="kc">None</span><span class="p">,</span>  <span class="kc">None</span><span class="p">,</span>  <span class="kc">None</span><span class="p">,</span>  <span class="kc">None</span><span class="p">,</span>  <span class="kc">None</span><span class="p">,</span>  <span class="kc">None</span><span class="p">],</span>
<span class="p">]</span>

<span class="c1"># Build DataFrame and save to CSV — this is the input format Scimap expects</span>
<span class="n">phenotype_workflow</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="n">columns</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Phenotyping workflow:&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">phenotype_workflow</span><span class="o">.</span><span class="n">to_string</span><span class="p">(</span><span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">))</span>

<span class="c1"># Save the workflow for reuse</span>
<span class="n">phenotype_workflow</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="s1">&#39;phenotype_workflow.csv&#39;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="run-the-classification">
<h3>Run the Classification<a class="headerlink" href="#run-the-classification" title="Permalink to this heading">#</a></h3>
<p>With our rules defined, we can now apply them. Scimap first <strong>rescales</strong> each marker’s intensity within [0, 1] per image (to correct for batch effects between cores), then applies the phenotyping logic.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Rescale marker intensities per image to correct for batch effects</span>
<span class="c1"># This maps each marker&#39;s distribution to [0, 1] so thresholds are comparable</span>
<span class="n">tma8</span> <span class="o">=</span> <span class="n">sm</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">rescale</span><span class="p">(</span><span class="n">tma8</span><span class="p">,</span> <span class="n">gate</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
<span class="n">tma1</span> <span class="o">=</span> <span class="n">sm</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">rescale</span><span class="p">(</span><span class="n">tma1</span><span class="p">,</span> <span class="n">gate</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Rescaling complete.&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Step 2: Apply the phenotyping rules to classify each cell</span>
<span class="n">phenotype_workflow</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;phenotype_workflow.csv&#39;</span><span class="p">)</span>

<span class="n">tma8</span> <span class="o">=</span> <span class="n">sm</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">phenotype_cells</span><span class="p">(</span><span class="n">tma8</span><span class="p">,</span> <span class="n">phenotype</span><span class="o">=</span><span class="n">phenotype_workflow</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;phenotype&#39;</span><span class="p">)</span>
<span class="n">tma1</span> <span class="o">=</span> <span class="n">sm</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">phenotype_cells</span><span class="p">(</span><span class="n">tma1</span><span class="p">,</span> <span class="n">phenotype</span><span class="o">=</span><span class="n">phenotype_workflow</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;phenotype&#39;</span><span class="p">)</span>

<span class="c1"># Checkpoint: How many cells were assigned to each phenotype?</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;=== Core 84 — Phenotype Counts ===&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">tma8</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;phenotype&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">value_counts</span><span class="p">()</span><span class="o">.</span><span class="n">to_string</span><span class="p">())</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">=== Core 42 — Phenotype Counts ===&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">tma1</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;phenotype&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">value_counts</span><span class="p">()</span><span class="o">.</span><span class="n">to_string</span><span class="p">())</span>
</pre></div>
</div>
</div>
</div>
</section>
</section>
<section id="step-4-quality-control">
<h2>Step 4: Quality Control<a class="headerlink" href="#step-4-quality-control" title="Permalink to this heading">#</a></h2>
<p><strong>Never skip QC.</strong> After automated phenotyping, we must verify the results make biological sense. We use two complementary checks:</p>
<p><strong>A) Marker heatmap</strong> — Each row is a phenotype; each column is a marker. If the logic is correct:</p>
<ul class="simple">
<li><p>The <code class="docutils literal notranslate"><span class="pre">Cancer</span></code> row should show high expression <strong>only</strong> for CK7 and ECadherin</p></li>
<li><p>The <code class="docutils literal notranslate"><span class="pre">Lymphoid</span></code> row should show high expression <strong>only</strong> for CD3d</p></li>
<li><p>If a phenotype row lights up for unexpected markers, your rules need refinement</p></li>
</ul>
<p><strong>B) Spatial scatter plot</strong> — Overlay the phenotype labels back onto the tissue to check that the spatial distribution makes anatomical sense (e.g., cancer cells should form coherent nests, not scatter randomly).</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># QC A: Marker expression heatmap — Core 84</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Heatmap for Core 84:&quot;</span><span class="p">)</span>
<span class="n">sm</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">heatmap</span><span class="p">(</span><span class="n">tma8</span><span class="p">,</span> <span class="n">groupBy</span><span class="o">=</span><span class="s1">&#39;phenotype&#39;</span><span class="p">,</span> <span class="n">subsetMarkers</span><span class="o">=</span><span class="n">marker_cols</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">5</span><span class="p">))</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># QC A: Marker expression heatmap — Core 42</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Heatmap for Core 42:&quot;</span><span class="p">)</span>
<span class="n">sm</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">heatmap</span><span class="p">(</span><span class="n">tma1</span><span class="p">,</span> <span class="n">groupBy</span><span class="o">=</span><span class="s1">&#39;phenotype&#39;</span><span class="p">,</span> <span class="n">subsetMarkers</span><span class="o">=</span><span class="n">marker_cols</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">5</span><span class="p">))</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># QC B: Spatial scatter plot of phenotypes — Core 84</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Spatial phenotype map for Core 84:&quot;</span><span class="p">)</span>
<span class="n">sm</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">spatial_scatterPlot</span><span class="p">(</span>
    <span class="n">tma8</span><span class="p">,</span> <span class="n">colorBy</span><span class="o">=</span><span class="s1">&#39;phenotype&#39;</span><span class="p">,</span>
    <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">6</span><span class="p">,</span> <span class="mi">5</span><span class="p">),</span> <span class="n">plotLegend</span><span class="o">=</span><span class="kc">True</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># QC B: Spatial scatter plot of phenotypes — Core 42</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Spatial phenotype map for Core 42:&quot;</span><span class="p">)</span>
<span class="n">sm</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">spatial_scatterPlot</span><span class="p">(</span>
    <span class="n">tma1</span><span class="p">,</span> <span class="n">colorBy</span><span class="o">=</span><span class="s1">&#39;phenotype&#39;</span><span class="p">,</span>
    <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">6</span><span class="p">,</span> <span class="mi">5</span><span class="p">),</span> <span class="n">plotLegend</span><span class="o">=</span><span class="kc">True</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="tip admonition">
<p class="admonition-title">QUESTIONS:</p>
<p>Examine the heatmaps and spatial plots above:</p>
<ol class="arabic simple">
<li><p>Do any phenotype rows in the heatmap show unexpectedly high expression for off-target markers? What might cause this?</p></li>
<li><p>Which phenotype is <strong>most sensitive</strong> to the threshold choice, and why? <em>(Hint: think about markers that are expressed broadly at low levels.)</em></p></li>
<li><p>Do the “Cancer” cells form coherent spatial nests in the tissue plot, or are they scattered? What does that tell you about the classification quality?</p></li>
</ol>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Clean up: Merge low-confidence &quot;likely-&quot; subcategories into &#39;Unknown&#39;</span>
<span class="c1"># These are cells that partially match a phenotype but don&#39;t meet all criteria</span>
<span class="k">for</span> <span class="n">tma</span> <span class="ow">in</span> <span class="p">[</span><span class="n">tma8</span><span class="p">,</span> <span class="n">tma1</span><span class="p">]:</span>
    <span class="n">tma</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;phenotype&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">tma</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;phenotype&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">replace</span><span class="p">({</span>
        <span class="s1">&#39;likely-Cancer&#39;</span><span class="p">:</span> <span class="s1">&#39;Unknown&#39;</span><span class="p">,</span>
        <span class="s1">&#39;likely-Stromal&#39;</span><span class="p">:</span> <span class="s1">&#39;Unknown&#39;</span>
    <span class="p">})</span>

<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Low-confidence cells re-labelled as &#39;Unknown&#39;.&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">Updated counts — Core 84:&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">tma8</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;phenotype&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">value_counts</span><span class="p">()</span><span class="o">.</span><span class="n">to_string</span><span class="p">())</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">Updated counts — Core 42:&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">tma1</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;phenotype&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">value_counts</span><span class="p">()</span><span class="o">.</span><span class="n">to_string</span><span class="p">())</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="step-5-measuring-the-social-distance-between-cell-types">
<h2>Step 5: Measuring the “Social Distance” Between Cell Types<a class="headerlink" href="#step-5-measuring-the-social-distance-between-cell-types" title="Permalink to this heading">#</a></h2>
<p>Now we get to the heart of the exercise. We will measure the <strong>physical distance</strong> from every Cancer cell to its nearest neighbour of each other phenotype.</p>
<section id="why-does-this-matter">
<h3>Why does this matter?<a class="headerlink" href="#why-does-this-matter" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>A <strong>left-shifted</strong> distance distribution (peak at short distances) means immune cells are <strong>physically close</strong> to tumour cells — consistent with active infiltration.</p></li>
<li><p>A <strong>right-shifted</strong> distribution (peak at longer distances) suggests immune cells are present in the tissue but <strong>excluded</strong> from the tumour.</p></li>
</ul>
<div class="admonition-before-running-the-code admonition">
<p class="admonition-title"><strong>Before running the code:</strong></p>
<p>Predict which core you think will show greater immune–tumour proximity, based on the spatial maps you examined in Step 4.</p>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Define a consistent colour palette for cell types across all plots</span>
<span class="n">celltype_colors</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s1">&#39;Cancer&#39;</span><span class="p">:</span>   <span class="s1">&#39;#D33F49&#39;</span><span class="p">,</span>  <span class="c1"># Red</span>
    <span class="s1">&#39;Stromal&#39;</span><span class="p">:</span>  <span class="s1">&#39;#FED18C&#39;</span><span class="p">,</span>  <span class="c1"># Peach</span>
    <span class="s1">&#39;Lymphoid&#39;</span><span class="p">:</span> <span class="s1">&#39;#6599CD&#39;</span><span class="p">,</span>  <span class="c1"># Blue</span>
    <span class="s1">&#39;Myeloid&#39;</span><span class="p">:</span>  <span class="s1">&#39;#508791&#39;</span><span class="p">,</span>  <span class="c1"># Teal</span>
    <span class="s1">&#39;Unknown&#39;</span><span class="p">:</span>  <span class="s1">&#39;#D9DEE4&#39;</span><span class="p">,</span>  <span class="c1"># Grey</span>
<span class="p">}</span>

<span class="c1"># Combine both cores into one AnnData object for side-by-side comparison</span>
<span class="n">adata</span> <span class="o">=</span> <span class="n">ad</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span>
    <span class="p">[</span><span class="n">tma8</span><span class="p">,</span> <span class="n">tma1</span><span class="p">],</span> <span class="n">join</span><span class="o">=</span><span class="s1">&#39;outer&#39;</span><span class="p">,</span>
    <span class="n">label</span><span class="o">=</span><span class="s1">&#39;batch&#39;</span><span class="p">,</span> <span class="n">keys</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;core84&#39;</span><span class="p">,</span> <span class="s1">&#39;core42&#39;</span><span class="p">]</span>
<span class="p">)</span>

<span class="c1"># Calculate pairwise spatial distances between all phenotype combinations</span>
<span class="n">adata</span> <span class="o">=</span> <span class="n">sm</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">spatial_distance</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">phenotype</span><span class="o">=</span><span class="s1">&#39;phenotype&#39;</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Spatial distances calculated. Summary:&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;  Total cells: </span><span class="si">{</span><span class="n">adata</span><span class="o">.</span><span class="n">n_obs</span><span class="si">:</span><span class="s2">,</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;  Phenotypes: </span><span class="si">{</span><span class="nb">list</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;phenotype&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">())</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Plot A: Numeric distance summary — median distance from Cancer to each cell type</span>
<span class="c1"># This gives a quick overview of the &#39;social landscape&#39; around tumour cells</span>
<span class="n">sm</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">spatial_distance</span><span class="p">(</span>
    <span class="n">adata</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s1">&#39;numeric&#39;</span><span class="p">,</span>
    <span class="n">distance_from</span><span class="o">=</span><span class="s1">&#39;Cancer&#39;</span><span class="p">,</span>
    <span class="n">log</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">height</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="n">aspect</span><span class="o">=</span><span class="mi">11</span><span class="o">/</span><span class="mi">8</span><span class="p">,</span>
    <span class="n">palette</span><span class="o">=</span><span class="n">celltype_colors</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Plot B: Distribution of Cancer → Lymphoid distances, comparing both cores</span>
<span class="c1"># This is the key diagnostic plot for immune infiltration vs exclusion</span>
<span class="n">sm</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">spatial_distance</span><span class="p">(</span>
    <span class="n">adata</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s1">&#39;distribution&#39;</span><span class="p">,</span>
    <span class="n">distance_from</span><span class="o">=</span><span class="s1">&#39;Cancer&#39;</span><span class="p">,</span> <span class="n">distance_to</span><span class="o">=</span><span class="s1">&#39;Lymphoid&#39;</span><span class="p">,</span>
    <span class="n">log</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">height</span><span class="o">=</span><span class="mi">4</span><span class="p">,</span> <span class="n">aspect</span><span class="o">=</span><span class="mf">1.5</span><span class="p">,</span>
    <span class="n">palette</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;TMA_8_core84&#39;</span><span class="p">:</span> <span class="s1">&#39;#FED18C&#39;</span><span class="p">,</span> <span class="s1">&#39;TMA_1_core42&#39;</span><span class="p">:</span> <span class="s1">&#39;#A7AEF2&#39;</span><span class="p">}</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="tip admonition">
<p class="admonition-title">TASK</p>
<p>Examine the Cancer → Lymphoid distance distribution above and answer:</p>
<ol class="arabic simple">
<li><p><strong>The “Close Contact” Peak:</strong> Which core has a higher density of Lymphoid cells at very short distances from Cancer cells?</p></li>
<li><p><strong>The “Exclusion” Gap:</strong> Does either core show a pronounced hump at a <em>longer</em> distance? This suggests immune cells are present in the tissue but physically barred from entering the tumour.</p></li>
<li><p><strong>Clinical Classification:</strong> In immunotherapy research, tumours are classified as <em>Inflamed</em> (T cells inside tumour), <em>Excluded</em> (T cells at border), or <em>Desert</em> (no T cells). Based only on this plot, how would you classify each core?</p></li>
</ol>
<blockquote>
<div><p>Did your prediction match the result? If not, what might explain the difference?</p>
</div></blockquote>
</div>
</section>
</section>
<section id="step-6-mapping-recurrent-cell-neighbourhoods-rcns">
<h2>Step 6: Mapping Recurrent Cell Neighbourhoods (RCNs)<a class="headerlink" href="#step-6-mapping-recurrent-cell-neighbourhoods-rcns" title="Permalink to this heading">#</a></h2>
<p>Distance analysis tells us <em>how far</em> cells are from each other. But to understand the full tissue architecture, we need to identify distinct <strong>Recurrent Cell Neighbourhoods (RCNs)</strong> — recurrent spatial patterns where specific combinations of cell types co-occur.</p>
<section id="the-rcn-approach-in-three-steps">
<h3>The RCN approach in three steps:<a class="headerlink" href="#the-rcn-approach-in-three-steps" title="Permalink to this heading">#</a></h3>
<ol class="arabic simple">
<li><p>For each cell, count how many neighbours of each phenotype live within a radius <em>r</em></p></li>
<li><p>Use <strong>k-means clustering</strong> on these neighbourhood profiles to group cells with similar local environments</p></li>
<li><p>Assign biological names to each cluster based on its dominant cell types</p></li>
</ol>
<div class="admonition-predict-before-you-compute admonition">
<p class="admonition-title">Predict before you compute</p>
<p>Before running the code, think: <strong>what distinct neighbourhoods would you expect to find in a tumour microenvironment?</strong> For example:</p>
<ul class="simple">
<li><p>Pure tumour cores?</p></li>
<li><p>Stromal regions?</p></li>
<li><p>Immune-rich zones?</p></li>
<li><p>Mixed interface regions?</p></li>
</ul>
<p>Write down your predictions, then compare them to the clusters you discover.</p>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Remove cells with ambiguous phenotype for cleaner neighbourhood analysis</span>
<span class="n">adata_known</span> <span class="o">=</span> <span class="n">adata</span><span class="p">[</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;phenotype&#39;</span><span class="p">]</span> <span class="o">!=</span> <span class="s1">&#39;Unknown&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>

<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Cells retained after removing Unknowns: </span><span class="si">{</span><span class="n">adata_known</span><span class="o">.</span><span class="n">n_obs</span><span class="si">:</span><span class="s2">,</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Phenotype breakdown:&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">adata_known</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;phenotype&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">value_counts</span><span class="p">()</span><span class="o">.</span><span class="n">to_string</span><span class="p">())</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># For every cell: count neighbours of each phenotype within a 50-pixel radius</span>
<span class="c1"># This radius is a biological decision — too small = noisy, too large = loses local structure</span>
<span class="n">sm</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">spatial_count</span><span class="p">(</span><span class="n">adata_known</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s1">&#39;radius&#39;</span><span class="p">,</span> <span class="n">radius</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;spatial_count&#39;</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Neighbourhood counts computed. Each cell now has a &#39;neighbourhood profile&#39;.&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="finding-the-optimal-number-of-neighbourhoods">
<h3>Finding the Optimal Number of Neighbourhoods<a class="headerlink" href="#finding-the-optimal-number-of-neighbourhoods" title="Permalink to this heading">#</a></h3>
<p>We use the <strong>elbow method</strong> to choose how many RCNs to define. We plot the k-means inertia (sum of squared distances to cluster centres) for different values of <em>k</em>. The “elbow” — where the inertia starts decreasing much more slowly — is a good choice for <em>k</em>.</p>
<blockquote>
<div><p><strong>What to look for:</strong> Find the value of <em>k</em> where the curve bends most sharply. Adding clusters beyond that point gives diminishing returns.</p>
</div></blockquote>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">sklearn.cluster</span><span class="w"> </span><span class="kn">import</span> <span class="n">KMeans</span>

<span class="c1"># Extract the neighbourhood count matrix</span>
<span class="n">neighbor_data</span> <span class="o">=</span> <span class="n">adata_known</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s1">&#39;spatial_count&#39;</span><span class="p">]</span>

<span class="c1"># Run k-means for k = 1 to 10, recording inertia at each step</span>
<span class="n">inertia</span> <span class="o">=</span> <span class="p">[]</span>
<span class="n">K_range</span> <span class="o">=</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">11</span><span class="p">)</span>

<span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="n">K_range</span><span class="p">:</span>
    <span class="n">km</span> <span class="o">=</span> <span class="n">KMeans</span><span class="p">(</span><span class="n">n_clusters</span><span class="o">=</span><span class="n">k</span><span class="p">,</span> <span class="n">random_state</span><span class="o">=</span><span class="mi">42</span><span class="p">,</span> <span class="n">n_init</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span>
    <span class="n">km</span><span class="o">.</span><span class="n">fit</span><span class="p">(</span><span class="n">neighbor_data</span><span class="p">)</span>
    <span class="n">inertia</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">km</span><span class="o">.</span><span class="n">inertia_</span><span class="p">)</span>

<span class="c1"># Plot the elbow curve</span>
<span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">5</span><span class="p">))</span>
<span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">K_range</span><span class="p">,</span> <span class="n">inertia</span><span class="p">,</span> <span class="s1">&#39;o-&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;#508791&#39;</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">markersize</span><span class="o">=</span><span class="mi">8</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s1">&#39;Number of Clusters (k)&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">13</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s1">&#39;Inertia (Within-Cluster Sum of Squares)&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">13</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;Elbow Method — Choosing the Number of RCNs&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">14</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">axvline</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;#D33F49&#39;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;--&#39;</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.7</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;Chosen k = 5&#39;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">legend</span><span class="p">()</span>
<span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>

<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">Based on the elbow curve, k=5 is a reasonable choice.&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Cluster cells into 5 Regional Cell Neighbourhoods</span>
<span class="c1"># Clustering both cores together ensures cluster labels are comparable across images</span>
<span class="n">sm</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">spatial_cluster</span><span class="p">(</span><span class="n">adata_known</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;spatial_cluster&#39;</span><span class="p">,</span> <span class="n">k</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s1">&#39;kmeans&#39;</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Cells assigned to 5 spatial clusters.&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">adata_known</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;spatial_cluster&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">value_counts</span><span class="p">()</span><span class="o">.</span><span class="n">to_string</span><span class="p">())</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="interpreting-the-neighbourhoods">
<h3>Interpreting the Neighbourhoods<a class="headerlink" href="#interpreting-the-neighbourhoods" title="Permalink to this heading">#</a></h3>
<p>Now we inspect the cell-type composition of each cluster to assign biological names. We use a stacked bar plot — the dominant phenotype in each cluster determines its identity.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Generate stacked barplots for each core and retrieve the underlying data</span>

<span class="kn">import</span><span class="w"> </span><span class="nn">tempfile</span>

<span class="k">with</span> <span class="n">tempfile</span><span class="o">.</span><span class="n">TemporaryDirectory</span><span class="p">()</span> <span class="k">as</span> <span class="n">tmpdir</span><span class="p">:</span>
    <span class="n">df_84</span> <span class="o">=</span> <span class="n">sm</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">stacked_barplot</span><span class="p">(</span>
        <span class="n">adata_known</span><span class="p">[</span><span class="n">adata_known</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;imageid&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;TMA_8_core84&#39;</span><span class="p">],</span>
        <span class="n">x_axis</span><span class="o">=</span><span class="s1">&#39;spatial_cluster&#39;</span><span class="p">,</span> <span class="n">y_axis</span><span class="o">=</span><span class="s1">&#39;phenotype&#39;</span><span class="p">,</span> <span class="n">return_data</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
        <span class="n">saveDir</span><span class="o">=</span><span class="n">tmpdir</span><span class="p">,</span> <span class="n">fileName</span><span class="o">=</span><span class="s1">&#39;tmp.pdf&#39;</span>
    <span class="p">)</span>

    <span class="n">df_42</span> <span class="o">=</span> <span class="n">sm</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">stacked_barplot</span><span class="p">(</span>
        <span class="n">adata_known</span><span class="p">[</span><span class="n">adata_known</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;imageid&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;TMA_1_core42&#39;</span><span class="p">],</span>
        <span class="n">x_axis</span><span class="o">=</span><span class="s1">&#39;spatial_cluster&#39;</span><span class="p">,</span> <span class="n">y_axis</span><span class="o">=</span><span class="s1">&#39;phenotype&#39;</span><span class="p">,</span> <span class="n">return_data</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
        <span class="n">saveDir</span><span class="o">=</span><span class="n">tmpdir</span><span class="p">,</span> <span class="n">fileName</span><span class="o">=</span><span class="s1">&#39;tmp.pdf&#39;</span>
    <span class="p">)</span>

<span class="c1"># Manually plot with the biological colour palette for consistency</span>
<span class="n">fig</span><span class="p">,</span> <span class="p">(</span><span class="n">ax1</span><span class="p">,</span> <span class="n">ax2</span><span class="p">)</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">16</span><span class="p">,</span> <span class="mi">6</span><span class="p">),</span> <span class="n">sharey</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

<span class="n">df_84</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">kind</span><span class="o">=</span><span class="s1">&#39;bar&#39;</span><span class="p">,</span> <span class="n">stacked</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">celltype_colors</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax1</span><span class="p">,</span> <span class="n">legend</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="n">ax1</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;Core 84: Neighbourhood Composition&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">13</span><span class="p">)</span>
<span class="n">ax1</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s1">&#39;Proportion of Cells&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">12</span><span class="p">)</span>
<span class="n">ax1</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s1">&#39;Spatial Cluster&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">12</span><span class="p">)</span>
<span class="n">ax1</span><span class="o">.</span><span class="n">tick_params</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="s1">&#39;x&#39;</span><span class="p">,</span> <span class="n">rotation</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>

<span class="n">df_42</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">kind</span><span class="o">=</span><span class="s1">&#39;bar&#39;</span><span class="p">,</span> <span class="n">stacked</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">celltype_colors</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax2</span><span class="p">)</span>
<span class="n">ax2</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;Core 42: Neighbourhood Composition&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">13</span><span class="p">)</span>
<span class="n">ax2</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s1">&#39;Spatial Cluster&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">12</span><span class="p">)</span>
<span class="n">ax2</span><span class="o">.</span><span class="n">tick_params</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="s1">&#39;x&#39;</span><span class="p">,</span> <span class="n">rotation</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="n">ax2</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">title</span><span class="o">=</span><span class="s1">&#39;Phenotype&#39;</span><span class="p">,</span> <span class="n">bbox_to_anchor</span><span class="o">=</span><span class="p">(</span><span class="mf">1.05</span><span class="p">,</span> <span class="mi">1</span><span class="p">),</span> <span class="n">loc</span><span class="o">=</span><span class="s1">&#39;upper left&#39;</span><span class="p">)</span>

<span class="n">plt</span><span class="o">.</span><span class="n">suptitle</span><span class="p">(</span><span class="s1">&#39;Regional Cell Neighbourhood Composition by Core&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">15</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="mf">1.02</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Print the raw proportions for detailed inspection</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;=== Core 84 — Cluster Compositions ===&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">df_84</span><span class="o">.</span><span class="n">round</span><span class="p">(</span><span class="mi">3</span><span class="p">)</span><span class="o">.</span><span class="n">to_string</span><span class="p">())</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">=== Core 42 — Cluster Compositions ===&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">df_42</span><span class="o">.</span><span class="n">round</span><span class="p">(</span><span class="mi">3</span><span class="p">)</span><span class="o">.</span><span class="n">to_string</span><span class="p">())</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="assign-biological-names-to-clusters">
<h3>Assign Biological Names to Clusters<a class="headerlink" href="#assign-biological-names-to-clusters" title="Permalink to this heading">#</a></h3>
<p>Based on the bar plots above, assign a biological name to each cluster. The template below uses example names — <strong>update the <code class="docutils literal notranslate"><span class="pre">rcn_names</span></code> dictionary to match what you observe in your own results</strong>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Update these IDs based on your actual stacked barplot output above</span>
<span class="c1"># The key is the cluster number (as a string); the value is the biological name</span>
<span class="n">rcn_names</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s1">&#39;0&#39;</span><span class="p">:</span> <span class="s1">&#39;Myeloid-Rich Niche&#39;</span><span class="p">,</span>
    <span class="s1">&#39;1&#39;</span><span class="p">:</span> <span class="s1">&#39;Stroma&#39;</span><span class="p">,</span>
    <span class="s1">&#39;2&#39;</span><span class="p">:</span> <span class="s1">&#39;Solid Tumour Core&#39;</span><span class="p">,</span>
    <span class="s1">&#39;3&#39;</span><span class="p">:</span> <span class="s1">&#39;Lymphoid-Rich Niche&#39;</span><span class="p">,</span>
    <span class="s1">&#39;4&#39;</span><span class="p">:</span> <span class="s1">&#39;Tumour-Stromal Interface&#39;</span>
<span class="p">}</span>

<span class="c1"># Add a new column with readable names, preserving the numeric cluster IDs</span>
<span class="n">adata_known</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;spatial_community&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span>
    <span class="n">adata_known</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;spatial_cluster&#39;</span><span class="p">]</span>
    <span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span>
    <span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="n">rcn_names</span><span class="p">)</span>
<span class="p">)</span>

<span class="c1"># Define a palette for the community names</span>
<span class="n">community_palette</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s1">&#39;Tumour-Stromal Interface&#39;</span><span class="p">:</span> <span class="s1">&#39;#508791&#39;</span><span class="p">,</span>  <span class="c1"># Teal</span>
    <span class="s1">&#39;Solid Tumour Core&#39;</span><span class="p">:</span>        <span class="s1">&#39;#DC19DC&#39;</span><span class="p">,</span>  <span class="c1"># Magenta</span>
    <span class="s1">&#39;Stroma&#39;</span><span class="p">:</span>                   <span class="s1">&#39;#FED18C&#39;</span><span class="p">,</span>  <span class="c1"># Peach</span>
    <span class="s1">&#39;Myeloid-Rich Niche&#39;</span><span class="p">:</span>       <span class="s1">&#39;#FA3232&#39;</span><span class="p">,</span>  <span class="c1"># Red</span>
    <span class="s1">&#39;Lymphoid-Rich Niche&#39;</span><span class="p">:</span>      <span class="s1">&#39;#A7AEF2&#39;</span><span class="p">,</span>  <span class="c1"># Lavender</span>
<span class="p">}</span>

<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Communities assigned:&quot;</span><span class="p">,</span> <span class="nb">list</span><span class="p">(</span><span class="n">adata_known</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;spatial_community&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">()))</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Final spatial maps coloured by Regional Cell Neighbourhood</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;RCN map — Core 84:&quot;</span><span class="p">)</span>
<span class="n">sm</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">spatial_scatterPlot</span><span class="p">(</span>
    <span class="n">adata_known</span><span class="p">[</span><span class="n">adata_known</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;imageid&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;TMA_8_core84&#39;</span><span class="p">],</span>
    <span class="n">colorBy</span><span class="o">=</span><span class="s1">&#39;spatial_community&#39;</span><span class="p">,</span> <span class="n">imageid</span><span class="o">=</span><span class="s1">&#39;imageid&#39;</span><span class="p">,</span>
    <span class="n">s</span><span class="o">=</span><span class="mi">12</span><span class="p">,</span> <span class="n">customColors</span><span class="o">=</span><span class="n">community_palette</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">5</span><span class="p">)</span>
<span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">RCN map — Core 42:&quot;</span><span class="p">)</span>
<span class="n">sm</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">spatial_scatterPlot</span><span class="p">(</span>
    <span class="n">adata_known</span><span class="p">[</span><span class="n">adata_known</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;imageid&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;TMA_1_core42&#39;</span><span class="p">],</span>
    <span class="n">colorBy</span><span class="o">=</span><span class="s1">&#39;spatial_community&#39;</span><span class="p">,</span> <span class="n">imageid</span><span class="o">=</span><span class="s1">&#39;imageid&#39;</span><span class="p">,</span>
    <span class="n">s</span><span class="o">=</span><span class="mi">12</span><span class="p">,</span> <span class="n">customColors</span><span class="o">=</span><span class="n">community_palette</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">5</span><span class="p">)</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="tip admonition">
<p class="admonition-title">TASK:</p>
<p>Examine the RCN spatial maps and consider:</p>
<ol class="arabic simple">
<li><p><strong>Core comparison:</strong> Which RCN is proportionally larger in Core 84 vs Core 42? How does this relate to the distance analysis in Step 5?</p></li>
<li><p><strong>Radius sensitivity:</strong> If you <strong>doubled</strong> the neighbourhood radius from 50 to 100 pixels, what would happen to the RCNs? Which biological structures would become <em>more</em> visible, and which would become <em>less specific</em>?</p></li>
</ol>
</div>
</section>
</section>
<section id="summary-conclusions">
<h2>Summary &amp; Conclusions<a class="headerlink" href="#summary-conclusions" title="Permalink to this heading">#</a></h2>
<p>In this exercise you applied a complete spatial proteomics analysis pipeline to two TMA cores:</p>
<div class="pst-scrollable-table-container"><table class="table">
<thead>
<tr class="row-odd"><th class="head"><p>Step</p></th>
<th class="head"><p>What you did</p></th>
<th class="head"><p>Key tool</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>1</p></td>
<td><p>Loaded and split the AnnData dataset</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">sc.read_h5ad</span></code></p></td>
</tr>
<tr class="row-odd"><td><p>2</p></td>
<td><p>Visualised raw marker expression spatially</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">sm.pl.spatial_scatterPlot</span></code></p></td>
</tr>
<tr class="row-even"><td><p>3</p></td>
<td><p>Built a rule-based cell phenotyping workflow</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">sm.tl.phenotype_cells</span></code></p></td>
</tr>
<tr class="row-odd"><td><p>4</p></td>
<td><p>Validated phenotypes with heatmaps and spatial plots</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">sm.pl.heatmap</span></code></p></td>
</tr>
<tr class="row-even"><td><p>5</p></td>
<td><p>Quantified immune–tumour proximity distributions</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">sm.tl.spatial_distance</span></code></p></td>
</tr>
<tr class="row-odd"><td><p>6</p></td>
<td><p>Discovered and named Recurrent Cell Neighbourhoods</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">sm.tl.spatial_count</span></code>, <code class="docutils literal notranslate"><span class="pre">sm.tl.spatial_cluster</span></code></p></td>
</tr>
</tbody>
</table>
</div>
<section id="key-takeaways">
<h3>Key Takeaways<a class="headerlink" href="#key-takeaways" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p><strong>Spatial context is everything.</strong> The same number of immune cells can have very different clinical relevance depending on <em>where</em> they are relative to the tumour.</p></li>
<li><p><strong>Rule-based phenotyping requires validation.</strong> Always sanity-check your classifications with heatmaps and spatial overlays before moving to downstream analysis.</p></li>
<li><p><strong>RCNs are scale-dependent.</strong> The radius parameter is a biological decision, not just a technical one — it determines the spatial scale at which you define “neighbourhood.”</p></li>
</ul>
</section>
</section>
<hr class="docutils" />
<section id="going-further">
<h2>Going Further<a class="headerlink" href="#going-further" title="Permalink to this heading">#</a></h2>
<p>Interested in extending this analysis? Here are some follow-up resources:</p>
<ul class="simple">
<li><p><a class="reference external" href="https://scimap.xyz">Scimap documentation</a></p></li>
<li><p><a class="reference external" href="https://anndata.readthedocs.io">AnnData documentation</a></p></li>
</ul>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            name: "python3",
            path: "./."
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

                </article>
              

              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="exercise2.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">Exercise 2: Biological interpretation &amp; cell phenotyping</p>
      </div>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#clinical-context">Clinical Context</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#what-you-will-discover">What You Will Discover</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#learning-outcomes">Learning outcomes</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#workflow-overview">Workflow Overview</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#step-1-load-the-data">Step 1: Load the Data</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#step-2-visualise-raw-marker-expression">Step 2: Visualise Raw Marker Expression</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#marker-reference-table">Marker Reference Table</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#step-3-define-cell-identities">Step 3: Define Cell Identities</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#the-rules">The Rules</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#run-the-classification">Run the Classification</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#step-4-quality-control">Step 4: Quality Control</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#step-5-measuring-the-social-distance-between-cell-types">Step 5: Measuring the “Social Distance” Between Cell Types</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#why-does-this-matter">Why does this matter?</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#step-6-mapping-recurrent-cell-neighbourhoods-rcns">Step 6: Mapping Recurrent Cell Neighbourhoods (RCNs)</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#the-rcn-approach-in-three-steps">The RCN approach in three steps:</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#finding-the-optimal-number-of-neighbourhoods">Finding the Optimal Number of Neighbourhoods</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#interpreting-the-neighbourhoods">Interpreting the Neighbourhoods</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#assign-biological-names-to-clusters">Assign Biological Names to Clusters</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#summary-conclusions">Summary &amp; Conclusions</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#key-takeaways">Key Takeaways</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#going-further">Going Further</a></li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By Ada, Ziqi and Aleksandra
</p>

  </div>
  
  <div class="footer-item">
    

  <p class="copyright">
    
      © Copyright 2022.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b"></script>
<script src="_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>